DOC: fix sphinx errors due to np.emath references

jon-deng opened the pull request:
Docstrings for some ufuncs reference np.emath which raises a warning
when sphinx generates documentation. Replacing the reference with
np.lib.scimath should fix the error, I think.

Currently I'm trying to figure out why changes to `numpy/core/code_generators/ufunc_docstrings.py` aren't reflected in numpy docstrings so have opened a draft pull request.

This should help fix some of the warnings raised in #13114.

mattip commented:
Changes in python and C documentation are only reflected in the sphinx build after rebuilding numpy: `pip install .; cd docs; make html`. 

You can see the ci result [here](https://app.circleci.com/pipelines/github/numpy/numpy/12810/workflows/66356fc3-fca9-4478-8d51-2019ddf929cc/jobs/24918) if you open the "build devdocs w/ref warnings" fold. Comparing this build to a [random PR's run](https://app.circleci.com/pipelines/github/numpy/numpy/12803/workflows/77d087cb-685e-4ee4-8c93-6a925bc775fa/jobs/24911) the number of warnings dropped from 190 to 184. :+1: 

There is still a mention of 
```
:mod:`emath`
```
in [the RST documentation](https://github.com/numpy/numpy/blob/main/doc/source/reference/routines.emath.rst), could you take a look at that too? I think it should become 
```
``emath``
```
with no link. What do you think?

jon-deng commented:
Hi matt, thanks for the feedback.

I had a look at `doc/source/reference/routines.emath.rst` and I think changing to `:mod:numpy.emath` to just ``emath`` with no link would be good.

It seems that sphinx can't find the `np.emath` module but I'm not sure why. The documentation page for `emath` (https://numpy.org/doc/stable/reference/routines.emath.html) says that `emath` should be available after numpy is imported. Do you know if sphinx just searches for `emath` in the numpy package after importing it?

mattip commented:
> It seems that sphinx can't find the np.emath module

I am not sure which occurence this is referring to, but if you are using `np.emath`, I am pretty sure you need to change that to `numpy.emath`

jon-deng commented:
Okay, thanks. I added the small additional change to `doc/source/reference/routines.emath.rst` replacing `:mod:numpy.emath` to ``numpy.emath`` as a code block. Is there anything else I should change as part of this PR?

mattip commented:
Doesn't `numpy.emath` works as a link? It seems strange to refer to this as a preferred alias but then to not be able to use it.

rossbar commented:
>  It seems strange to refer to this as a preferred alias but then to not be able to use it.

Agreed - I'm -1 on changing all the links from `emath.*` to `lib.scimath.*` even though it fixes the broken links, since the docs state that `emath` is the preferred module name. I think the right way to fix this would be to update the module name/alias to reflect the "desired" access name. Not quite sure how difficult that will be but it's likely to be more involved than a docs-only change.

jon-deng committed:
DOC: Fix broken links to numpy.emath module

There are a number of broken links to the numpy.emath module because
there are no generated documentation for numpy.emath, which is an alias
of numpy.lib.scimath. This fixes the broken links by generating
documentation for numpy.emath, that duplicates the documentation for
numpy.lib.scimath.

jon-deng commented:
I agree it's a bit weird.

I figured out a better way to fix it I think. We can add two module documentation files `doc/source/reference/routines.lib.scimath.rst` and `doc/source/reference/routines.emath.rst` so each one generates the same documentation but with appropriate module reference. The documentation in `routines.lib.scimath.rst` is under the module `lib.scimath` while the documentation in `routines.emath.rst` is under the module `emath`.

jon-deng commented:
Do you have any ideas how to fix the ci/circleci errors? It says there's an error in the Sphinx parallel build but I can't think of why that would be.

rossbar commented:
> Do you have any ideas how to fix the ci/circleci errors? It says there's an error in the Sphinx parallel build but I can't think of why that would be.

I think this was a problem with one of the dependencies which has been fixed, try rebasing on (or merging with) `main`.

jon-deng committed:
Merge branch 'main' into fix-sphinx-doc-errors

jon-deng committed:
DOC: Remove routines.lib.scimath.rst

jon-deng committed:
DOC: Fix missing lib.scimath.sqrt function reference

jon-deng committed:
DOC: Fix missing lib.scimath module reference

jon-deng commented:
I think I've fixed the issue. I ended up changing the `routines.emath.rst` file to reference the `numpy.emath` module whereas previously `routines.emath.rst` referenced the `numpy.lib.scimath` module. This fixes the missing links to function belonging to `numpy.emath` and allows using the preferred alias.

mattip commented on doc/build-errors.txt:
```diff
@@ -0,0 +1,5815 @@
+# for testing
```
Committed by mistake?

mattip commented on doc/source/reference/routines.emath.rst:
```diff
@@ -1,9 +1,9 @@
-Mathematical functions with automatic domain (:mod:`numpy.emath`)
+Mathematical functions with automatic domain
 ***********************************************************************
```
I think the underline length does not match the text.

mattip commented on doc/source/reference/routines.emath.rst:
```diff
@@ -1,9 +1,9 @@
 .. currentmodule:: numpy
 
-.. note:: :mod:`numpy.emath` is a preferred alias for :mod:`numpy.lib.scimath`,
+.. note:: :mod:`numpy.emath` is a preferred alias for ``numpy.lib.scimath``,
```
revert this change?

jon-deng replied:
The link to `` :mod:`numpy.lib.scimath` `` is broken now because I changed the line at the bottom of this file from ``..automodule numpy.lib.scimath`` to ``..automodule numpy.emath``. Because of that, Sphinx generates references to members of `numpy.emath` which fixes the links but links to `numpy.lib.scimath` are now missing.

I was thinking of potentially adding the `numpy.lib.scimath` documentation but it seems there are no documents for the `numpy.lib.*` modules in `doc/source/reference`?  Because of that, I'm unsure if it should be documented/how to document it? What do you think?

I think there are a few other warnings (aside from `emath` related ones) also related to missing `numpy.lib.*` functions and modules.

mattip commented:
Nice. If this is ready please remove the draft status

jon-deng committed:
DOC: Remove accidentally added build-errors.txt file

jon-deng committed:
DOC: Fix formatting for routines.emath.rst title

InessaPawson commented:
@mattip It's now ready for review.

mattip commented:
Thanks @jon-deng

jon-deng commented:
No problem and thanks for merging.

InessaPawson commented:
Hi-five on merging your first pull request to NumPy, @jon-deng! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion
